stl/inc/flat_map

```diff
@@ -0,0 +1,1000 @@
+
+        using iterator_category = input_iterator_tag;
+        using iterator_concept  = random_access_iterator_tag;
+        using difference_type   = ptrdiff_t;
+        using value_type        = pair<iter_value_t<_Key_iterator_t>, iter_value_t<_Mapped_iterator_t>>;
+        using reference         = pair<iter_const_reference_t<_Key_iterator_t>, iter_reference_t<_Mapped_iterator_t>>;
+
+    private:
+        class _Arrow_proxy {
+        public:
+            explicit _Arrow_proxy(const reference& _Rx) noexcept : _Ref{_Rx} {}
+
+            const reference* operator->() const noexcept {
+                return _STD addressof(_Ref);
+            }
+
+        private:
+            reference _Ref;
+        };
+
+    public:
+        using pointer = _Arrow_proxy;
+
+        reference operator*() const {
+            return reference{*_Key_it, *_Mapped_it};
+        }
+
+        pointer operator->() const {
+            return pointer{*(*this)};
+        }
+
+        type& operator++() {
+            ++_Key_it;
+            ++_Mapped_it;
+            return *this;
+        }
+
+        type operator++(int) {
+            type _Tmp = *this;
+            ++*this;
+            return _Tmp;
+        }
+
+        bool operator==(const type& _Right) const {
+            return _Key_it == _Right._Key_it;
+        }
+
+        auto operator<=>(const type& _Right) const {
+            return _Key_it <=> _Right._Key_it;
+        }
+
+        type& operator--() {
+            --_Key_it;
+            --_Mapped_it;
+            return *this;
+        }
+
+        type operator--(int) {
+            type _Tmp = *this;
+            --*this;
+            return _Tmp;
+        }
+
+        type& operator+=(difference_type _Off) {
+            _Key_it += _Off;
+            _Mapped_it += _Off;
+            return *this;
+        }
+
+        type& operator-=(difference_type _Off) {
+            _Key_it -= _Off;
+            _Mapped_it -= _Off;
+            return *this;
+        }
+
+        type operator+(difference_type _Off) const {
+            type _Tmp = *this;
+            return _Tmp += _Off;
+        }
+
+        type operator-(difference_type _Off) const {
+            type _Tmp = *this;
+            return _Tmp -= _Off;
+        }
+
+        reference operator[](difference_type _Off) const {
+            return *(*this + _Off);
+        }
+
+        difference_type operator-(const type& _Right) const {
+            return _Key_it - _Right._Key_it;
+        }
+
+        friend type operator+(difference_type _Off, const type& _Right) {
+            return _Right + _Off;
+        }
+
+        template <bool _Val = _IsConst, typename = enable_if_t<!_Val, int>>
+        operator typename _Flat_map_iterator_Impl<enable_if_t<!_Val, _KeyContainer>,
+            enable_if_t<!_Val, _MappedContainer>, true>::type() const {
+            return typename _Flat_map_iterator_Impl<_KeyContainer, _MappedContainer, true>::type{_Key_it, _Mapped_it};
+        }

```

Comment:
This conversion function looks weird. Given we're already requiring concepts, can we constrain it with `requires (!_IsConst)`?

Reply:
I tried but I couldn't get it to work. If you can let me know how to make the fix, then I'll definitely apply it.

Here is the code in my attempt:
```
        template <bool _Enable = !_IsConst> requires _Enable
        operator typename _Flat_map_iterator_Impl<_KeyContainer, _MappedContainer, true>::type() const {
            return typename _Flat_map_iterator_Impl<_KeyContainer, _MappedContainer, true>::type{_Key_it, _Mapped_it};
        }
```
clang compiler complains that `conversion function converting 'std::_Flat_map_iterator_Impl<std::vector<int>, std::vector<int>, true>::type' to itself will never be used [-Werror,-Wclass-conversion]`, meaning that the constraint is not disabling the function in `_Flat_map_iterator_Impl<..., false>` correctly.

Reply:
It seems to me that we can just write the following:
```C++
        operator typename _Flat_map_iterator_Impl<_KeyContainer>, _MappedContainer>, true>::type() const requires (!_IsConst) {
            return typename _Flat_map_iterator_Impl<_KeyContainer, _MappedContainer, true>::type{_Key_it, _Mapped_it};
        }
```
[Godbolt link](https://godbolt.org/z/s81o5T91W) for a reduced example.

Reply:
Fixed.